README.md
# A PyTorch implementation of a YOLO v3 Object Detector

[UPDATE] : This repo serves as a driver code for my research. I just graduated college, and am very busy looking for research internship / fellowship roles before eventually applying for a masters. I won't have the time to look into issues for the time being. Thank you.


This repository contains code for a object detector based on [YOLOv3: An Incremental Improvement](https://pjreddie.com/media/files/papers/YOLOv3.pdf), implementedin PyTorch. The code is based on the official code of [YOLO v3](https://github.com/pjreddie/darknet), as well as a PyTorch 
port of the original code, by [marvis](https://github.com/marvis/pytorch-yolo2). One of the goals of this code is to improve
upon the original port by removing redundant parts of the code (The official code is basically a fully blown deep learning 
library, and includes stuff like sequence models, which are not used in YOLO). I've also tried to keep the code minimal, and 
document it as well as I can. 

### Tutorial for building this detector from scratch
If you want to understand how to implement this detector by yourself from scratch, then you can go through this very detailed 5-part tutorial series I wrote on Paperspace. Perfect for someone who wants to move from beginner to intermediate pytorch skills. 

[Implement YOLO v3 from scratch](https://blog.paperspace.com/how-to-implement-a-yolo-object-detector-in-pytorch/)

As of now, the code only contains the detection module, but you should expect the training module soon. :) 

## Requirements
1. Python 3.5
2. OpenCV
3. PyTorch 0.4
4. Tensorflow
... see the Req. Document
Using PyTorch 0.3 will break the detector.



## Detection Example + Calculating DISTANCE

![Detection Example](https://i.imgur.com/ros7bw3.jpg)



### Speed Accuracy Tradeoff
You can change the resolutions of the input image by the `--reso` flag. The default value is 416. Whatever value you chose, rememeber **it should be a multiple of 32 and greater than 32**. Weird things will happen if you don't. You've been warned. 

```
python detect.py --images imgs --det det --reso 320
```

### On Video
For this, you should run the file, video_demo.py with --video flag specifying the video file. The video file should be in .avi format
since openCV only accepts OpenCV as the input format. 

```
python video_demo.py --video video.avi
```

Tweakable settings can be seen with -h flag. 


### Get Started
Same as video module, but you don't have to specify the video file since feed will be taken from your camera. To be precise, 
feed will be taken from what the OpenCV, recognises as camera 0. The default image resolution is 160 here, though you can change it with `reso` flag.

```
python main.py

or

python maintest.py --video videotest.mp4
```



pyaudio/CHANGELOG
2017-03-18  Hubert Pham  <hubert@mit.edu>

  PyAudio 0.2.11

  - Fix use-after-free memory issue in callback handler.

    Thanks to both Blaise Potard and Matthias Schaff for their patches!

  - Fix docstring for get_output_latency().

    Thanks to Timothy Port for finding the issue!

2017-01-10  Hubert Pham  <hubert@mit.edu>

  PyAudio 0.2.10

  - Release the GIL during PortAudio I/O calls to avoid potential deadlock.

    Thanks to Michael Graczyk for submitting a patch!

  - Add a few automated unit tests.

2015-10-18  Hubert Pham  <hubert@mit.edu>

  PyAudio 0.2.9

  - Fix overflow error handling logic for pa_read_stream.

    Stream.read takes an additional parameter that specifies whether
    an exception is raised on audio buffer overflow, for parity with
    Stream.write.  Includes relevant bug fixes in the C module logic.

    Thanks to Tony Jacobson for submitting a patch!

  - Fix IOError arguments.

    IOError exceptions previously had values in the strerror and errno fields
    swapped, which is now corrected.

    Thanks to Sami Liedes for the report!

  - Miscellaneous updates.

    Python library surfaces issues with importing low-level C module.
    Code formatting update.
    Updates to examples for Python 3 compatibility.

2014-02-16  Hubert Pham  <hubert@mit.edu>

  PyAudio 0.2.8

  - Device names: support non-UTF8 encoded device names.

    get_device_info_by_index() now attempts to decode the device name
    using a few known encodings (defaults to UTF-8 and CP1252).  If
    those fail, PyAudio passes the raw bytes for the device name.
    Previously, PyAudio assumed a UTF-8 encoding, which is not always
    true.

  - Callback-mode: fix deadlock on some platforms when calling
    pa.stop_stream.

    Thanks to Jason Roehm for this patch!

2012-10-20  Hubert Pham  <hubert@mit.edu>

  PyAudio 0.2.7

  - Callback-mode: support callables.

    Thanks to John Luebs and Bastian Bechtold for this patch.

  - Update documentation to use Sphinx.

    Thanks again to Bastian Bechtold for his incredible contribution!


2012-09-01  Hubert Pham  <hubert@mit.edu>

  PyAudio 0.2.6

  - Added support for Python 3.  As of this update, PyAudio is
    compatible with Python 2.6, Python 2.7, and Python 3.2.

    Many thanks to Bastian Bechtold and Bob Jamison for their patches!

  - Fixed a bug in which a list could be modified during iteration.

    Many thanks to Danilo J. S. Bellini for reporting this error!

  - Fixed a memory bug involving Mac OS X channel maps.


2012-09-01  Hubert Pham  <hubert@mit.edu>

  PyAudio 0.2.5

  - Added support for callback (non-blocking) operation.

    Many thanks to Bastian Bechtold for his initial contribution and
    his generous help towards releasing this feature.  Callback mode
    would not have happened without Bastian's help!


2010-08-12  Hubert Pham  <hubert@mit.edu>

  PyAudio 0.2.4

  - Maintenance release: updated directory structure and packaging.


2008-10-29  Hubert Pham  <hubert@mit.edu>

  PyAudio 0.2.3

  - Release the GIL during blocking PortAudio I/O calls.

  - Fixed Python argument parsing to use a long for PaSampleFormat
    (rather than int).

    Thanks to many who have pointed out these two issues and sent
    patches.

  - pyaudio.PyAudio.is_format_supported() now throws a ValueError
    exception if the specified format is not supported for any reason
    (or returns True if the format is supported).

    Prior, the method would return False if and only if the specified
    sample rate was unsupported. is_format_supported() now will always
    return True or throw an exception.


2008-03-06  Hubert Pham <hubert@mit.edu>

  PyAudio 0.2.0

  - Added PaMacCoreStreamInfo for Mac OS X Host API Specific Stream Info
    (e.g., for channel maps).

  - Added packaging files for building binaries.


2008-02-12  Justin Mazzola Paluska  <jmp@mit.edu>

  - Initial version of debian packaging.


MainTEST.py
from __future__ import division
import time
import torch 
import torch.nn as nn
from torch.autograd import Variable
import numpy as np
import cv2 
from util import *
from darknet import Darknet
from preprocess import prep_image, inp_to_image, letterbox_image
import pandas as pd
import random 
import argparse
import pickle as pkl
import denegement, getvoice
from bs4 import BeautifulSoup
import requests
from selenium import webdriver
from threading import Thread
import pyautogui, time, os
import win32com.client as wincl


label_detected = []
counting = 0

def get_test_input(input_dim, CUDA):
    img = cv2.imread("dog-cycle-car.png")
    img = cv2.resize(img, (input_dim, input_dim)) 
    img_ =  img[:,:,::-1].transpose((2,0,1))
    img_ = img_[np.newaxis,:,:,:]/255.0
    img_ = torch.from_numpy(img_).float()
    img_ = Variable(img_)
    
    if CUDA:
        img_ = img_.cuda()
    
    return img_



def prep_image(img, inp_dim):
    """
    Prepare image for inputting to the neural network. 
    
    Returns a Variable 
    """

    orig_im = img
    dim = orig_im.shape[1], orig_im.shape[0]
    img = (letterbox_image(orig_im, (inp_dim, inp_dim)))
    img_ = img[:,:,::-1].transpose((2,0,1)).copy()
    img_ = torch.from_numpy(img_).float().div(255.0).unsqueeze(0)
    return img_, orig_im, dim

 
def distance_to_camera(knownWidth, focalLength, perWidth):
	# compute and return the distance from the maker to the camera
	return ((knownWidth * focalLength) / perWidth)  * 2.54 / 96

old_distance,temp_Y1,temp_Y2, temp_X1,temp_X2 = 0 , 0 ,0 ,0, 0


def write(x, img):
    global counting
    global temp_Y1,temp_Y2, temp_X1,temp_X2, old_distance
    c1 = tuple(x[1:3].int())
    c2 = tuple(x[3:5].int())
    ver = 0
    #get width and height
    x1 = str(c1[0]).split('(')[1].split(',')[0]
    x2 = str(c1[1]).split('(')[1].split(',')[0]
    y1 = str(c2[0]).split('(')[1].split(',')[0]
    y2 = str(c2[1]).split('(')[1].split(',')[0]
    X = int(y1)-int(x1)
    Y = int(y2)-int(x2)
    rotating_1 = int(y2)-int(x1)
    rotating_2 = int(x2)-int(y1)
    surface_old = (temp_Y1-temp_X1)*(temp_Y2-temp_X2)
    #Focal length of the camera
    F = 155* 4.35
    obj_width = 340.15
    obj_height = 1040.15
    cls = int(x[-1])
    label = "{0}".format(classes[cls])
    unit = ' cm'
    if X != 0 and ( str(label) == 'traffic light' or str(label) == 'stop sign' or str(label) == 'fire hydrant') :
        if old_distance * 0.9 < int(round(distance_to_camera(obj_height,F,X))) < old_distance * 1.1 and temp_X2 * 0.9 < int(x2) < temp_X2 * 1.1:
            print('standing still ',old_distance)
            counting +=1
            ver = 1
            if counting == 10 and old_distance <1000:
                TEXT = 'WARNING the distance between you and the '+str(label)+' is '+str(old_distance)+ unit+ 'you should stop immidiatly.'
                Thread(target = text_to_speech(TEXT)).start()
                counting = 0
        else:
            temp_ditance = round(distance_to_camera(obj_height,F,X))
            if temp_ditance > 1000:
                unit = ' m'
                temp_ditance = temp_ditance/100
                obj_height = obj_height/100
                print('the distance between the camera and  '+str(label)+' is ',temp_ditance, unit)
                TEXT ='the distance between the camera and  '+str(label)+' is '+str(temp_ditance)+ unit
                counting = 0
            elif temp_ditance < 110:
                print('WARNING the distance between you and the '+str(label)+' is ',temp_ditance, unit, 'you should stop immidiatly.')
                TEXT ='WARNING the distance between you and the '+str(label)+' is '+str(temp_ditance)+ unit+ 'you should stop immidiatly.'
                text_to_speech(TEXT) 
            
            
        #print(X, '   ' , Y)
        #print(x1, '   ' ,x2, '   ' ,y1, '   ' ,y2)
        #if ver == 0 and ((temp_X1 > int(x1) and temp_X2 < int(x2)) or (temp_Y2 < int(y2) and temp_Y1 > int(y1))):
            #if surface_old*0.9 < X*Y < surface_old*1.1 :
                #print('complex rotation')
            #else:
                #print('simple rotation ')

        temp_X1,temp_X2 = int(x1), int(x2)
        temp_Y1,temp_Y2 = int(y1), int(y2)
        old_distance = int(round(distance_to_camera(obj_height,F,X)))

    color = random.choice(colors)
    cv2.rectangle(img, c1, c2,color, 1)
    t_size = cv2.getTextSize(label, cv2.FONT_HERSHEY_PLAIN, 1 , 1)[0]
    c2 = c1[0] + t_size[0] + 3, c1[1] + t_size[1] + 4
    tem = cv2.rectangle(img, c1, c2,color, -1)
    cv2.rectangle(img, c1, c2,color, -1)
    cv2.putText(img, label, (c1[0], c1[1] + t_size[1] + 4), cv2.FONT_HERSHEY_PLAIN, 1, [225,255,255], 1);
    #sizesh =  img.shape
    #height, width = tem.shape[:2]
    return img

def arg_parse():
    """
    Parse arguements to the detect module
    
    """
    
    
    parser = argparse.ArgumentParser(description='YOLO v3 Video Detection Module')
   
    parser.add_argument("--video", dest = 'video', help = 
                        "Video to run detection upon",
                        default = "video5.mp4", type = str)
    parser.add_argument("--dataset", dest = "dataset", help = "Dataset on which the network has been trained", default = "pascal")
    parser.add_argument("--confidence", dest = "confidence", help = "Object Confidence to filter predictions", default = 0.5)
    parser.add_argument("--nms_thresh", dest = "nms_thresh", help = "NMS Threshhold", default = 0.4)
    parser.add_argument("--cfg", dest = 'cfgfile', help = 
                        "Config file",
                        default = "cfg/yolov3.cfg", type = str)
    parser.add_argument("--weights", dest = 'weightsfile', help = 
                        "weightsfile",
                        default = "yolov3.weights", type = str)
    parser.add_argument("--reso", dest = 'reso', help = 
                        "Input resolution of the network. Increase to increase accuracy. Decrease to increase speed",
                        default = "416", type = str)
    return parser.parse_args()

def text_to_speech(TEXT):
    
    speak = wincl.Dispatch("SAPI.SpVoice")

    if TEXT:
        speak.Speak(TEXT)
        TEXT = ''

    
    
def main_main():
    global classes, colors
    cfgfile = "cfg/yolov3.cfg"
    weightsfile = "yolov3.weights"
    num_classes = 80

    args = arg_parse()
    confidence = float(args.confidence)
    nms_thesh = float(args.nms_thresh)
    start = 0
    CUDA = torch.cuda.is_available()
    

    
    
    num_classes = 80
    bbox_attrs = 5 + num_classes
    
    model = Darknet(cfgfile)
    model.load_weights(weightsfile)
    
    model.net_info["height"] = args.reso
    inp_dim = int(model.net_info["height"])
    
    assert inp_dim % 32 == 0 
    assert inp_dim > 32

    if CUDA:
        model.cuda()

    model(get_test_input(inp_dim, CUDA), CUDA)
            
    model.eval()
    
    videofile = args.video
    
    cap = cv2.VideoCapture(videofile)
    
    assert cap.isOpened(), 'Cannot capture source'
    
    frames = 0
    start = time.time()    
    while cap.isOpened():
        
        ret, frame = cap.read()
        if ret:
            
            img, orig_im, dim = prep_image(frame, inp_dim)
            im_dim = torch.FloatTensor(dim).repeat(1,2)                        
            
            
            if CUDA:
                im_dim = im_dim.cuda()
                img = img.cuda()
            
            
            with torch.no_grad():   
                output = model(Variable(img), CUDA)
            output = write_results(output, confidence, num_classes, nms = True, nms_conf = nms_thesh)

            if type(output) == int:
                frames += 1
                #print("FPS of the video is {:5.2f}".format( frames / (time.time() - start)))
                cv2.imshow("frame", orig_im)
                key = cv2.waitKey(1)
                if key & 0xFF == ord('q'):
                    break
                continue
            

        
            im_dim = im_dim.repeat(output.size(0), 1)
            scaling_factor = torch.min(inp_dim/im_dim,1)[0].view(-1,1)
            
            output[:,[1,3]] -= (inp_dim - scaling_factor*im_dim[:,0].view(-1,1))/2
            output[:,[2,4]] -= (inp_dim - scaling_factor*im_dim[:,1].view(-1,1))/2
            
            output[:,1:5] /= scaling_factor
    
            for i in range(output.shape[0]):
                output[i, [1,3]] = torch.clamp(output[i, [1,3]], 0.0, im_dim[i,0])
                output[i, [2,4]] = torch.clamp(output[i, [2,4]], 0.0, im_dim[i,1])
            
            classes = load_classes('data/coco.names')
            colors = pkl.load(open("pallete", "rb"))
            
            list(map(lambda x: write(x, orig_im), output))
            
            
            cv2.imshow("frame", orig_im)
            key = cv2.waitKey(1)
            if key & 0xFF == ord('q'):
                break
            frames += 1
            #print("FPS of the video is {:5.2f}".format( frames / (time.time() - start)))

            
        else:
            break

        
def check_voice_comm():
    while True:
        voice = ''
        try:
            voice = getvoice.get_voice()
        except:
            voice = ''
        if 'station' in voice:
            denegement.display_ip()

    
if __name__ == '__main__':
    #denegement.display_ip()
    Thread(target = main_main).start()
    Thread(target = check_voice_comm).start()
    #Thread(target = text_to_speech).start()


    

